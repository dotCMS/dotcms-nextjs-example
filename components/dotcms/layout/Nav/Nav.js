import React, { useState, useEffect, useContext } from 'react';
import { useRouter } from 'next/router';

import LanguageSelector from '../LanguageSelector';
import MenuList from '../../layout/Nav/MenuList';
import PageContext from '../../../../contexts/PageContext';
import SocialMediaMenu from '../../layout/Nav/SocialMediaMenu';
import logo from '../../../../public/logo.png';
import menuIcon from '../../../../public/menu.svg';
import useNav from '../../../../hooks/useNav';
import { MainNav, NavMenu } from '../../../../styles/nav/nav.styles';
import RouterLink from '../../../RouterLink';
import { getCurrentLanguage } from '../../../../utilities/dotcms/locale';

export default function Nav() {
    const nav = useNav();
    const [isMenuOpen, setIsMenuOpen] = useState(false);

    const {
        languageProps: { defaultLanguage }
    } = useContext(PageContext);

    const handleOpenMenu = (e) => {
        e.preventDefault();
        setIsMenuOpen(!isMenuOpen);
    };

    return (
        <div className="container">
            <MainNav className="main-nav">
                <RouterLink
                    href={
                        getCurrentLanguage() && !defaultLanguage ? `/${getCurrentLanguage()}` : `/`
                    }
                >
                    <img src={logo} alt="" width={135} height={41} />
                </RouterLink>
                <NavMenu className="main-nav__menu" isOpen={isMenuOpen}>
                    <a
                        className="hamburger"
                        href="#"
                        aria-label="button"
                        onClick={(e) => handleOpenMenu(e)}
                    >
                        <img src={menuIcon} alt="Hamburger Icon" />
                    </a>
                    <MenuList navigation={nav} />
                    <SocialMediaMenu />
                    <LanguageSelector />
                </NavMenu>
            </MainNav>
        </div>
    );
}
