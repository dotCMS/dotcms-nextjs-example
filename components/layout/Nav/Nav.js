import ItemLink from './ItemLink';
import NavDropDown from './NavDropDown';
import NavBarSearch from './NavBarSearch';

import { PageContext } from '../../../pages/_app';

export default function Nav() {
    return (
        <div className="rd-navbar-main-element">
            <PageContext.Consumer>
                {({ nav }) => (
                    <>
                        <div className="rd-navbar-nav-wrap">
                            <ul className="rd-navbar-nav">
                                {nav.map((item) => {
                                    if (item.type === 'folder' && item?.children?.length) {
                                        return <NavDropDown key={item.hash} options={item} />;
                                    } else {
                                        return (
                                            <li className="rd-nav-item" key={item.hash}>
                                                <ItemLink
                                                    className="rd-nav-link"
                                                    pathname={item.href}
                                                >
                                                    {item.title}
                                                </ItemLink>
                                            </li>
                                        );
                                    }
                                })}
                            </ul>
                        </div>
                        <NavBarSearch />
                    </>
                )}
            </PageContext.Consumer>
        </div>
    );
}
