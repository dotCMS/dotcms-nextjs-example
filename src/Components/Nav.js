import React, { Component } from 'react';
import { Nav as BootstrapNav, NavItem, NavLink, NavbarToggler, Collapse } from 'reactstrap';
import { Link } from 'react-router-dom';

export default class Nav extends Component {
    state = {
        isOpen: false,
        items: []
    };

    async componentDidMount() {
        fetch('/api/v1/nav//?depth=2', {
            headers: {
                DOTAUTH: Buffer.from(
                    `${process.env.REACT_APP_USER_EMAIL}:${process.env.REACT_APP_USER_PASSWORD}`
                ).toString('base64')
            }
        })
            .then(response => response.json())
            .then(data => data.entity.children)
            .then(data => {
                this.setState({
                    ...this.state,
                    items: data
                });
            });
    }

    toggleOpen = e => {
        this.setState({
            ...this.state,
            isOpen: !this.state.isOpen
        });
    };

    render() {
        return (
            <>
                <NavbarToggler onClick={this.toggleOpen} />
                <Collapse isOpen={this.state.isOpen} navbar>
                    <BootstrapNav pills={true}>
                        {this.state.items.map(item => {
                            return (
                                <NavItem key={item.folder}>
                                    <NavLink tag={Link} active={item.href === window.location.pathname} to={item.href}>
                                        {item.title}
                                    </NavLink>
                                </NavItem>
                            );
                        })}
                    </BootstrapNav>
                </Collapse>
            </>
        );
    }
}
