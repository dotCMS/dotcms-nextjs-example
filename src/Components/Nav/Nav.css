div.nav-menu__links {
    flex-grow: unset;
}

button.nav-menu__toggle-button.navbar-toggler {
    color: var(--white);
    border-color: transparent;
}
