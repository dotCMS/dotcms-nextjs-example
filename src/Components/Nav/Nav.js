import React, { Component } from 'react';
import { Nav as BootstrapNav, NavbarToggler, Collapse } from 'reactstrap';
import DotCMSApi from '../../libs/dotcms.api';
import PageContext from '../../PageContext';
import NavOption from '../../Components/Nav/NavOption';
import NavDropDown from './NavDropDown';
import NavSort from './NavSort';
import './Nav.css';

export default class Nav extends Component {
    static contextType = PageContext;
    state = {
        collapsed: false,
        showLogin: false,
        items: []
    };

    componentDidMount() {
        DotCMSApi.request({
            url: `${process.env.REACT_APP_DOTCMS_HOST}/api/v1/nav//?depth=3`
        })
            .then(response => response.json())
            .then(data => data.entity.children)
            .then(data => {
                this.setState({
                    ...this.state,
                    items: data
                });
            })
            .catch(e => console.error(`.catch(${e})`));
    }

    toggleCollapsed = e => {
        this.setState({
            ...this.state,
            collapsed: !this.state.collapsed
        });
    };

    toggleLogin = e => {
        this.setState({
            ...this.state,
            showLogin: !this.state.showLogin
        });
    };

    render() {
        const isEditModeFromDotCMS = this.context && this.context.mode && this.context.mode === 'EDIT_MODE';
        const remoteRendered = this.context && this.context.page && this.context.page.remoteRendered;

        return (
            <>
                {remoteRendered && isEditModeFromDotCMS ? (<NavSort></NavSort>) : ''}
                <NavbarToggler className="nav-menu__toggle-button" onClick={this.toggleCollapsed} >&#9776;</NavbarToggler>
                <Collapse className="nav-menu__links" isOpen={this.state.collapsed} navbar>
                    <BootstrapNav>
                        {this.state.items.map((item) => {
                            if (item.children.length && item.type === 'folder') {
                                return <NavDropDown key={item.hash} options={item} />;
                            } else {
                                return <NavOption key={item.hash} item={item} />;
                            }
                        })}
                    </BootstrapNav>
                </Collapse>
            </>
        );
    }
}
