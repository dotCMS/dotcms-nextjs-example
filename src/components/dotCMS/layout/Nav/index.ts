export * from './MenuList'
export * from './Nav'
export * from './NavBarSearch'
export * from './NavDropDown'
export * from './NavOption'
export * from './NavSort'
export * from './SocialMediaMenu'
