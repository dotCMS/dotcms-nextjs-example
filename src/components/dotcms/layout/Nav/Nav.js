import React, { useState, useContext } from 'react'
import Image from 'next/image'

import LanguageSelector from '../LanguageSelector'
import MenuList from './MenuList'
import PageContext from '../../../../contexts/PageContext'
import RouterLink from '../../../RouterLink'
import SocialMediaMenu from './SocialMediaMenu'
import useNav from '../../../../hooks/useNav'
import { MainNav, NavMenu } from '../../../../styles/nav/nav.styles'
import { getCurrentLanguage } from '../../../../utilities/dotcms/locale'

export default function Nav() {
  const nav = useNav()
  const [isMenuOpen, setIsMenuOpen] = useState(false)

  const {
    languageProps: { defaultLanguage },
  } = useContext(PageContext)

  const handleOpenMenu = (e) => {
    e.preventDefault()
    setIsMenuOpen(!isMenuOpen)
  }

  return (
    <div className="container">
      <MainNav className="main-nav">
        <RouterLink
          href={
            getCurrentLanguage() && !defaultLanguage
              ? `/${getCurrentLanguage()}`
              : `/`
          }
        >
          <Image
            alt="DotCMS - NextJS example"
            height="41"
            src="/logo.png"
            width="135"
          />
        </RouterLink>
        <NavMenu isOpen={isMenuOpen}>
          <button
            aria-label="button"
            className="hamburger"
            href="#"
            onClick={(e) => handleOpenMenu(e)}
          >
            <Image alt="Open Menu" height="24" src="/menu.svg" width="24" />
          </button>
          <MenuList navigation={nav} />
          <SocialMediaMenu />
          <LanguageSelector />
        </NavMenu>
      </MainNav>
    </div>
  )
}
