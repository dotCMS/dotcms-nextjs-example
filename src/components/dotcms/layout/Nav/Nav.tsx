// Dependencies
import * as React from 'react'
import Image from 'next/image'

// Internals
import { Link } from '@/components'
import { PageContext } from '@/contexts'
import { useNav } from '@/hooks'
import { getCurrentLanguage } from '@/lib/dotCMS'
import { MainNav, NavMenu } from '@/styles/nav/nav.styles'
import { LanguageSelector } from '../LanguageSelector'
import MenuList from './MenuList'
import SocialMediaMenu from './SocialMediaMenu'

export const Nav = () => {
  const nav = useNav()
  const [isMenuOpen, setIsMenuOpen] = React.useState(false)

  const {
    // @ts-ignore TODO: add the correct types
    languageProps: { defaultLanguage },
  } = React.useContext(PageContext)

  const handleOpenMenu = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault()
    setIsMenuOpen(!isMenuOpen)
  }

  return (
    <div className="container">
      <MainNav className="main-nav">
        <Link
          href={
            getCurrentLanguage() && !defaultLanguage
              ? `/${getCurrentLanguage()}`
              : `/`
          }
        >
          <Image
            alt="DotCMS - NextJS example"
            height="41"
            src="/logo.png"
            width="135"
          />
        </Link>
        <NavMenu isOpen={isMenuOpen}>
          <button
            aria-label="button"
            className="hamburger"
            onClick={(e) => handleOpenMenu(e)}
          >
            <Image alt="Open Menu" height="24" src="/menu.svg" width="24" />
          </button>
          <MenuList navigation={nav} />
          <SocialMediaMenu />
          <LanguageSelector />
        </NavMenu>
      </MainNav>
    </div>
  )
}

export default Nav
