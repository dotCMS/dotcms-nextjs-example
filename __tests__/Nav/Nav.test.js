import React from 'react';
import { shallow } from 'enzyme';
import wait from 'waait';
import Nav from '../../components/dotcms/layout/Nav/Nav';
import toJSON from 'enzyme-to-json';

describe('<Nav />', () => {
    let wrapper;

    beforeEach(async () => {
        wrapper = await shallow(<Nav />);
    });

    it('renders nav correctly', async () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('should expand menu', async () => {
        expect(wrapper.find('Collapse').props().isOpen).toBe(false);

        wrapper.find('NavbarToggler').simulate('click');
        await wrapper.update();
        await wait();

        expect(wrapper.find('Collapse').props().isOpen).toBe(true);
    });
});
