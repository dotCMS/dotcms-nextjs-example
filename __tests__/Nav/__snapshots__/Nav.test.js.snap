// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders nav correctly 1`] = `
<Fragment>
  <JSXStyle
    id="3435437959"
  >
    .nav-menu__links.navbar-collapse{float:right;}.navbar-expand-md .navbar-nav .dropdown-menu.show{position:relative;}.nav-menu__toggle-button.navbar-toggler{color:var(--white);border-color:transparent;}.nav-option{font-size:0.75rem;font-weight:500;-webkit-letter-spacing:2px;-moz-letter-spacing:2px;-ms-letter-spacing:2px;letter-spacing:2px;}.nav-option,.nav-option:focus,.nav-option:hover{color:var(--white);text-transform:uppercase;}.navbar-expand-md .nav-menu__links .nav-option.nav-link{padding-left:1rem;}.nav-menu__links .nav-option.nav-link.active{background-color:transparent;color:rgba(255,255,255,0.5);}.nav-menu__links.navbar-collapse ul{background-color:var(--gray);bottom:0;display:inline-block;left:0;min-width:208px;padding:0.5rem;position:fixed;top:0;}.nav-menu__links .nav-dropdown{background-color:var(--white);border:none;box-shadow:none;color:var(--black);float:none;margin:0 1rem 1rem;padding:0;position:static;}.nav-dropdown .dropdown-item{color:var(--black);font-size:0.8125rem;font-weight:500;line-height:1.2;padding:0.5rem;}.nav-dropdown .dropdown-item.active,.nav-dropdown .dropdown-item:hover{color:var(--white);}.nav-dropdown .dropdown-item:hover{background-color:var(--blue);color:var(--white);}.nav-sort__button{margin-right:0.5rem;}@media (max-width:992px) and (min-width:768px){.navbar-expand-md .nav-menu__links.navbar-collapse{display:none !important;}.navbar-expand-md .nav-menu__links.navbar-collapse.show{display:unset !important;}.nav-menu__toggle-button.navbar-toggler{display:block !important;}}@media screen and (min-width:992px){.nav-dropdown::before{content:'';border-bottom:solid 0.625rem var(--black);border-left:solid 0.625rem transparent;border-right:solid 0.625rem transparent;height:0.625rem;width:1.25rem;position:absolute;top:-9px;right:45px;}.nav-menu__links.navbar-collapse ul{background-color:transparent;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;padding:0;position:relative;}.navbar-expand-md .nav-menu__links.navbar-collapse{display:unset !important;}.navbar-expand-md .navbar-nav .dropdown-menu.show{position:absolute;}.nav-menu__links .nav-dropdown{background-color:var(--black);box-shadow:0px 14px 33px -9px rgba(0,0,0,0.75);min-width:unset;padding:1.25rem;}.nav-dropdown .dropdown-item{color:#999999;}.nav-dropdown .dropdown-item.active{background-color:transparent;color:var(--white);}.nav-dropdown .dropdown-item:hover{background-color:unset;color:var(--white);}}
  </JSXStyle>
  <ContextConsumer>
    <Component />
  </ContextConsumer>
</Fragment>
`;
